### DISCUSSION TITLE
How to translate the using statement of C# to PascalABC.NET?

### DISCUSSION THREAD

**Discussion by ghost** [pascalabcnet/pascalabcnet#2935] [Q&A] [answered]
The last time I worked with C#, there is only the using directive that equivalent to Java's import.

**Comment by SunSerega** [comment]
`using` only adds names from a specified namespace into context. In Pascal, it's replaced with `uses`.

So this code:
```
## uses System.Collections;

var a := new BitArray(16);
Println(a);
```
Produces the same .exe file as this:
```
##
var a := new System.Collections.BitArray(16);
Println(a);
```
(and you can still use full class names, even when the namespace is used)

But uses in Pascal can also refer to a .pas file:
```
uses Unit1;
uses Unit2 in '..\a\Unit2';
uses Other.Name; // "." (dot) cannot be in Pascal unit name, so this can only be a namespace
```

**Reply by ghost** [reply]
This is the using directive. What I'm talking about is the using statement.

**Reply by SunSerega** [reply, marked as answer]
Ah, right...
There is no equivalent, but in C# it is a syntax sugar that unwraps into something like:
```
begin
  var x := new DisposableObject;
  try
    // body
  finally
    System.IDisposable(x).Dispose();
  end;
end;
```

**Reply by ghost** [reply]
How to translate this piece of code to PascalABC.NET?

```
using (Lua lua = new Lua()) // Create the Lua script engine
{
    dynamic env = lua.CreateEnvironment(); // Create a environment
    env.dochunk("a = 'Hallo World!';", "test.lua"); // Create a variable in Lua
    Console.WriteLine(env.a); // Access a variable in C#
    env.dochunk("function add(b) return b + 3; end;", "test.lua"); // Create a function in Lua
    Console.WriteLine("Add(3) = {0}", env.add(3)); // Call the function in C#
}
```

My translation doesn't compile:

```
uses Neo.IronLua;

begin
  var lua := new Lua;
  try
    var env := lua.CreateEnvironment();
    env.dochunk('a = "Hallo World!";', 'test.lua');
    //writeln(env.a);
    env.dochunk('function add(b) return b + 3; end;', 'test.lua');
    //writeln('Add(3) = {0}', env.add(3));
  finally
    System.IDisposable(lua).Dispose();
  end;
end.
```

Note: I don't know how to escape the `'` character. I tried `\'` but the IDE's syntax highlighting broken, indicating that this is wrong. I finally choose to use `"` instead. The commented out lines are what causing compilation problems.

The first commented out line causing this compilation error:

`Project5.pas(8) : The type 'Neo.IronLua.LuaGlobal' does not contain a definition for 'a'`

The second commented out line causing this compilation error:

`Project5.pas(10) : Function expected, procedure 'Add' met`

I tried to remove `{0}` but it still gives the same error.

**Reply by SunSerega** [reply]
>  I don't know how to escape the `'` character.

```
$'The "''" character'.Println;
```
If you add it twice in a row - it adds once instead of ending the string.

---

As for `dynamic` - didn't get to use it in C#. But again, it's just a syntax sugar to write less. Should still be possible to do everything by just putting a dot after `env` on a new line and looking through what the code analyzer finds in that variable.

**Reply by ghost** [reply]
Both `env.dochunk($'a = 'Hallo World!';', 'test.lua');` and `env.dochunk($'a = "'Hallo World!'";', 'test.lua');` are invalid code. I want to pass Lua code as is to the method, not print it to the screen.

**Reply by ghost** [reply]
`env.dochunk($'a = "'"Hallo World!"'";', 'test.lua');` is also invalid code.

**Reply by SunSerega** [reply]
Because they are not defined in whatever type `lua.CreateEnvironment` has returned. Instead, they are defined as dynamic attachments, which you can get/call using some methods, which IDE shows you when you start typing `env` and then place a dot after it.

**Reply by SunSerega** [reply]
Ah, right, but they are invalid because you didn't double the `'` symbol. Again, I said you need to write it twice in a row. If you actually launch the last code line I sent - you'll see that the `"` symbols left as is, but the doubled `'` symbol has undoubled itself in the output.

**Reply by ghost** [reply]
I give up. It seems the PascalABC.NET compiler will not allow me to use NeoLua. Thank you anyway.

**Reply by SunSerega** [reply]
I'm pretty sure there is nothing impossible here... But I'm even less invested in the topic than you, so whatever.

**Reply by ghost** [reply]
No. I think it's impossible without making changes to the compiler.

**Reply by samuraiGH** [reply]
most likely you just don’t know PascalABC.NET well

you don't even provide error codes

**Reply by ghost** [reply]
> most likely you just don’t know PascalABC.NET well
> 
> you don't even provide error codes

Yes, I don't know PascalABC.NET well. But I did provide error codes: https://github.com/pascalabcnet/pascalabcnet/discussions/2935#discussioncomment-7329347
